docs/proposals/drafts/20211007-pluggable-authenticators.md

```diff
@@ -0,0 +1,143 @@
+
+## Implementation Options
+
+1. Importing for side-effect
+2. Go Plugin Module
+3. OAuth2 Providers
+
+## 1. Importing for Side-Effects
+In this approach, importing for Go's side-effects is used to enable each authenticator to register itself through a map of authenticator types and their factories.
+
+To onboard a new authenticator, the following is needed:
+
+1. Implement the corresponding interface.
+2. Implement a factory.
+3. Register the factory in the authenticators' factories map.
+
+With this approach, an example tenant configuration would look like:
+
+```
+- name: test-oidc
+  id: 1610b0c3-c509-4592-a256-a1871353dbfa
+  authenticator:
+    type: oidc
+    config:
+      clientID: test
+      clientSecret: xyz
+      issuerCAPath: ./tmp/certs/ca.pem
+      issuerURL: http://127.0.0.1:5556/dex
+      redirectURL: https://localhost:8443/oidc/test-oidc/callback
+      usernameClaim: email
+  rateLimits:
+    - endpoint: "/api/metrics/v1/.+/api/v1/receive"
+      limit: 100
+      window: 1s
+    - endpoint: "/api/logs/v1/.*"
+      limit: 100
+      window: 1s
+```
+
+## 2. Go Plugin Module
+
+In this approach, a generic interface for Authentication providers (Authenticators) is introduced. Authentication providers then implement that interface in a Go plugin. This plugin will be built and loaded at run time by the Observatorium API. How and where to load the plugin module is defined by the tenants configuration file.
+
+This approach is easy to implement, very flexible in terms of adding more authenticators without any additional implementation and integration complexity. However, there are few concerns when it comes to trust and security, since the Observatorium API would consume a pre-built plugin binary as an authenticator.
+
+**Reference**: [Writing Modular Go Programs with Plugins](https://medium.com/learning-the-go-programming-language/writing-modular-go-programs-with-plugins-ec46381ee1a9)
+
+## 3. OAuth2 Providers
+
+Similar to the [generic OAuth2 providers in grafana](https://github.com/grafana/grafana/blob/main/pkg/login/social/social.go), OAuth2 providers can be introduced and injected into the supported OAuth2 alongside to the current implementation of OIDC and mTLS to prevent any disruption for the current adopters. While this approach provides flexibility and can be considered to be an elegant and clean way of introducing more authenticators support, it requires a lot of changes in the Observatorium API.
```

Comment:
```suggestion
Similar to the [generic OAuth2 providers in grafana](https://github.com/grafana/grafana/blob/main/pkg/login/social/social.go), OAuth2 providers can be introduced and injected into the supported OAuth2 alongside the current implementation of OIDC and mTLS to prevent any disruption for the current adopters. While this approach provides flexibility and can be considered to be an elegant and clean way of introducing more authenticators support, it requires a lot of changes in the Observatorium API.
```